### ISSUE TITLE
JITLink: __eh_frame emission/JIT EH frame registration on aarch64-darwin

### ISSUE THREAD

**Issue by dnadlinger** [llvm/llvm-project#52921]
To be able to `unw_step` through JIT-ed frames on aarch64-darwin, it is necessary to emit __eh_frame unwind information, as there is no way to register compact unwind info with libunwind.

However, JITLink drops all that unwind information on MachO-arm64, as the `EHFrame{Splitter, EdgeFixer}` passes are only added to the default link graph on x86_64. This effectively makes `EHFrameRegistrationPlugin` a no-op.

Just adding the passes the same way as on x86_64 appears to work, although I haven't checked whether there are any extra relocations to support, …:

```diff
commit 57eb8c10025bdbca8f8260bd43a46d0aca369fe8 (HEAD -> julia-main, dnadlinger/julia-main)
Author: David Nadlinger <code@klickverbot.at>
Date:   Thu Dec 30 04:23:23 2021 +0100

    [JITLink] Also include EHFrame* passes on MachO_arm64

    Without the non-compact unwind info there is no way to register
    the EH frames with libunwind (e.g. for backtrace generation).

diff --git a/llvm/lib/ExecutionEngine/JITLink/MachO_arm64.cpp b/llvm/lib/ExecutionEngine/JITLink/MachO_arm64.cpp
index f2a029d35cd5..4d958b302ff9 100644
--- a/llvm/lib/ExecutionEngine/JITLink/MachO_arm64.cpp
+++ b/llvm/lib/ExecutionEngine/JITLink/MachO_arm64.cpp
@@ -705,6 +705,10 @@ void link_MachO_arm64(std::unique_ptr<LinkGraph> G,
     Config.PrePrunePasses.push_back(
         CompactUnwindSplitter("__LD,__compact_unwind"));

+    Config.PrePrunePasses.push_back(EHFrameSplitter("__TEXT,__eh_frame"));
+    Config.PrePrunePasses.push_back(EHFrameEdgeFixer("__TEXT,__eh_frame",
+        8, Delta64, Delta32, NegDelta32));
+
     // Add an in-place GOT/Stubs pass.
     Config.PostPrunePasses.push_back(
         PerGraphGOTAndPLTStubsBuilder_MachO_arm64::asPass);
```

(There is also the somewhat related, non-JITLink-specific issue of __eh_frame sometimes being elided, for which Julia currently carries a patch: https://github.com/JuliaLang/llvm-project/commit/60e041894288848e37870c42749a1aabcc2c2274.)

**Comment by dnadlinger** [comment]
(@lhames would probably be the person to ping.)

**Comment by lhames** [comment]
@dnadlinger LGTM -- I can commit this to the mainline on your behalf. Do you have a preferred name/email that I can use for commit?

**Comment by dnadlinger** [comment]
@lhames: Ah, great, thanks, then I won't open a Phabricator ticket – I just wasn't sure whether this was the correct approach, or I was missing something entirely. `David Nadlinger <code@klickverbot.at>` should be fine.

**Comment by lhames** [comment]
@dnadlinger Phab or an llvm-commits email is standard practice for submitting, but this patch is short and sweet so there's no need to submit an extra review here.

**Comment by lhames** [comment]
Sorry for the delay on this -- I need to write up a testcase. I hope to get one done and land this tomorrow.

**Comment by lhames** [comment]
No luck yet -- my simple testcase (see below) is crashing under `llvm-jitlink -orc-runtime=libclang_rt.orc_ios.a testcase.o` in `unw_set_reg`. I believe there is an outstanding bug for this -- I'll look into that in a day or two.

```
int main(int argc, char *argv[]) {
  try {
    throw 42;
  } catch (int x) {
    return x;
  }
  return 0;
}
```

**Comment by lhames** [comment]
> (There is also the somewhat related, non-JITLink-specific issue of __eh_frame sometimes being elided, for which Julia currently carries a patch: JuliaLang@60e0418.)

Ahh, I should have read more carefully. Ok -- there's no easy way to generate a testcase in-tree at the moment. I'm going to commit this without a test case and include a FIXME explaining the issue (the goal is to support compact unwind registration in JITLink and libunwind in the near future).

**Comment by lhames** [comment]
@dnadlinger -- thanks again for this. 253ce92 should take care of this issue.

Side note: I wasn't able to reproduce the `unw_set_reg` crash today. I think there are still latent bugs here, but since they're already being tracked by https://llvm.org/PR49692 there's no reason to keep this issue open for them.

**Comment by nico** [comment]
253ce92844f72e3a6d0e423473f2765c2c5afd6a breaks tests on Windows: http://45.33.8.238/win/52383/summary.html (just the ExecutionEngine/JITLink/AArch64/MachO_arm64_ehframe.s failure, the other failing test is a different regression.)

Please take a look and revert for now if it takes a while to fix.

**Comment by nico** [comment]
Reverted in 66b2ed477f902329c916bd4389b2034dae09c02b for now.

**Comment by lhames** [comment]
This patch was re-applied with fixes in bfd0cbd4eb9fef85da96bb799038ba49a52d08f8. It appears to be passing all tests now.
